optee-utee/src/tee_parameter.rs

```diff
@@ -0,0 +1,302 @@
+
+    pub fn written_slice(&self) -> Option<&[u8]> {
+        match &self.content {
+            ParamContent::MemrefOutput { buffer, written } => Some(&buffer[..*written]),
+            ParamContent::MemrefInout { buffer, written } => Some(&buffer[..*written]),
+            _ => None,
+        }
+    }
+
+    pub fn output_value(&self) -> Option<(u32, u32)> {
+        match &self.content {
+            ParamContent::ValueOutput { a, b } => Some((*a, *b)),
+            ParamContent::ValueInout { a, b } => Some((*a, *b)),
+            _ => None,
+        }
+    }
+
+    fn get_type(&self) -> ParamType {
+        match &self.content {
+            ParamContent::None => ParamType::None,
+            ParamContent::MemrefInput { .. } => ParamType::MemrefInput,
+            ParamContent::MemrefOutput { .. } => ParamType::MemrefOutput,
+            ParamContent::MemrefInout { .. } => ParamType::MemrefInout,
+            ParamContent::ValueInput { .. } => ParamType::ValueInput,
+            ParamContent::ValueOutput { .. } => ParamType::ValueOutput,
+            ParamContent::ValueInout { .. } => ParamType::ValueInout,
+        }
+    }
+
+    fn get_raw_type(&self) -> u32 {
+        self.get_type() as u32
+    }
+
+    fn as_raw(&mut self) -> raw::TEE_Param {
+        match &mut self.content {
+            ParamContent::None => raw::TEE_Param {
+                memref: raw::Memref {
+                    buffer: core::ptr::null_mut(),
+                    size: 0,
+                },
+            },
+            ParamContent::MemrefInput { buffer } => raw::TEE_Param {
+                memref: raw::Memref {
+                    buffer: (*buffer).as_ptr() as *mut core::ffi::c_void,
+                    size: buffer.len(),
+                },
+            },
+            ParamContent::MemrefOutput { buffer, written: _ } => raw::TEE_Param {
+                memref: raw::Memref {
+                    buffer: (*buffer).as_mut_ptr() as *mut core::ffi::c_void,
+                    size: buffer.len(),
+                },
+            },
+            ParamContent::MemrefInout { buffer, written: _ } => raw::TEE_Param {
+                memref: raw::Memref {
+                    buffer: (*buffer).as_mut_ptr() as *mut core::ffi::c_void,
+                    size: buffer.len(),
+                },
+            },
+            ParamContent::ValueInput { a, b } => raw::TEE_Param {
+                value: raw::Value { a: *a, b: *b },
+            },
+            ParamContent::ValueInout { a, b } => raw::TEE_Param {
+                value: raw::Value { a: *a, b: *b },
+            },
+            ParamContent::ValueOutput { a, b } => raw::TEE_Param {
+                value: raw::Value { a: *a, b: *b },
+            },
+        }
+    }
+
+    fn update_size_from_raw(&mut self, raw_param: &raw::TEE_Param) {
+        match &mut self.content {
+            ParamContent::MemrefOutput { buffer: _, written } => {
+                *written = unsafe { raw_param.memref.size };
```

Comment:
I am curious why we consider accessing `raw::TEE_Param` is `unsafe` here. Can you add comment about the reason to the code?

Reply:
Accessing `union` should be unsafe in Rust:
error[E0133]: access to union field is unsafe and requires unsafe function or block
   --> /teaclave/optee-utee/src/tee_parameter.rs:152:28
    |
152 |                 *written = raw_param.memref.size;
    |                            ^^^^^^^^^^^^^^^^ access to union field
    |
    = note: the field may not be properly initialized: using uninitialized data will cause undefined behavior

Reply:
I suggest we follow `https://github.com/rust-lang/rust-clippy/issues/9330` to explicitly add comment to `unsafe` block; and run `cargo lint` to help automatically enforce available rules.

Reply:
Will add `SAFETY` comment for the modules inside this PR. Since `cargo clippy` infects many other modules, there will be another PR to address all rules.